### PR TITLE
Question Service with CRUD and Filter Operations

### PR DESCRIPTION
**Pull request by elaineshijie**
Addresses #26 , #27, #28 

### Question Service

Created a new question service that performs CRUD and filter operations. Implemented validation with Joi but yet to add tests. Please set up your own .env file with DEV_PORT (8003) & DEV_URI. 
- Added ACCESS_TOKEN_SECRET to .env for verification of accessToken

#### Added endpoints:
- POST /question-service: Create a new question. [admin user only]
- GET /question-service: Retrieve a list of questions
- GET /question-service/:id: Retrieve a specific question by ID.
- PUT /question-service/:id: Update an existing question. [admin user only] 
   - does not require all attributes to be present (intended behaviour: will not throw error if attributes are not changed)
- DELETE /question-service/:id: Delete a question by ID. [admin user only]
- GET /question-service/filter/:  Filter questions based on provided query parameters (category & difficulty string arrays that are optional).
   - if no category / difficulty specified, by default all categories/ difficulty levels are accepted
- GET /question-service/random/:  Find a random question matching provided query parameters (category & difficulty optional arrays).
   - if no category / difficulty specified, by default all categories/ difficulty levels are accepted
- GET  /question-service/category/: Get all the question categories in existing (not deleted) questions

#### Question criteria
- Questions can have the same title as long as they have different difficulty levels. (POST & PUT)
- Questions cannot have the same description at all. (POST & PUT)
- Question description is an object.
- Question categories have to be one or more with no restrictions (case-insensitive); will store & return as uppercase
- Question difficulty is restricted to easy, medium, hard (case-insensitive); will store & return as uppercase
- All fields are required.

#### Other implementation details
- Soft delete: deleting a question and then creating the same question is allowed.
- Implemented CORS middleware to allow requests from specified origins (port number 8000, 8001, 8002, 8004 local host & frontend dev, prod and staging)
- **Decryption of accessToken to verify isAdmin done in question service for now**  - POST, PUT, DELETE
- [TODO: verify isAdmin field in cookie]

#### Schema updates
- no image stored for each question (stored together with description object)
- add testCases in Question (refer to schema in `model.js`)
  - testCase: meta is set as mixed type & optional; all other fields are required

### CODE REVIEW

File: question-service/middlewares/cors.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+const allowedOrigins = [
+  "http://localhost:8000",
+  "http://localhost:8001",
+  "http://localhost:8002",
+  "http://localhost:8004",
+];
+
+// PORT 8000 - FRONTEND
+// PORT 8001 - AUTH SERVICE
```
Comment (pzl111) [inline question-service/middlewares/cors.js:11]: minor nit i think its user service not auth service

File: question-service/router/router.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import express from "express";
+import { createQuestion, getAllQuestions, getQuestionById, deleteQuestionById, updateQuestionById, getFilteredQuestions, findQuestion } from "../manager/questionManager.js";
+import validateQuestion from '../middlewares/validation.js';
+
+const router = express.Router();
+
+// CREATE NEW QUESTION
+router.post('/', validateQuestion, createQuestion);
```
Comment (pzl111) [inline question-service/router/router.js:8]: is this route meant to be protected? (i.e. only accessible by admins)

File: question-service/router/router.js [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+router.route("/").get(getAllQuestions);
+
+// GET QUESTION BY ID
+router.route("/id/:id").get(getQuestionById);
+
+// DELETE QUESTION BY ID
+router.route("/id/:id").delete(deleteQuestionById);
+
+// UPDATE QUESTION BY ID
+router.route("/id/:id").put(updateQuestionById);
```
Comment (pzl111) [inline question-service/router/router.js:20]: should create, delete and update be accessible only to admins? (while get is anyone)

File: question-service/middlewares/cors.js (hunk 1 above)
Comment (pzl111) [inline question-service/middlewares/cors.js:11]: minor nit line 11 i think its user service not auth service

File: question-service/index.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+
+dotenv.config();
+const app = express();
+const port = process.env.DEV_PORT || 8003;
+
+
+app.use(corsMiddleware);
+app.use(loggingMiddleware);
+app.use(bodyParser.json());
```
Comment (pzl111) [inline question-service/index.js:18]: i think you might also need cookieParser middleware to parse cookies from HTTP requests

**Review by pzl111 — CHANGES_REQUESTED** [review]
Some minor comments

File: question-service/index.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+app.use("/api/question-service", router);
+
+app.use(errorHandler);
+
+// Test Route for Health Checks
+app.get("/healthz", (req, res) => {
+  res.status(200).json({ message: "Connected to /healthz route of question-service" });
+});
```
Comment (georgetayqy) [inline question-service/index.js:29]: Added a healthz path for health checks needed by AWS ALB.

File: question-service/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+EXPOSE 8003
+
+# copy all files into the Docker container
+COPY . .
+
+# install node libraries
+RUN npm install
+
+# execute the server
+CMD ["node", "index.js"]
```
Comment (georgetayqy) [inline question-service/Dockerfile:14]: Added a Dockerfile to help dockerise the application.

**Review by georgetayqy — APPROVED** [review]
LGTM!

The service is deployed onto AWS and is available at the following links:

API Gateway: https://n4qffw5pr5.execute-api.ap-southeast-1.amazonaws.com/v1/api/question-service
Application Load Balancer: Hidden to prevent authorised access
ECS: Hidden behind Application Load Balancer

**Review by pzl111 — APPROVED** [review]

File: question-service/middlewares/cors.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  "http://localhost:5173", // frontend dev
+  "http://peerprep.s3-website-ap-southeast-1.amazonaws.com", // frontend prod
+  "http://peerprep-frontend-bucket.s3-website-ap-southeast-1.amazonaws.com", // frontend staging
+  "http://localhost:8000",
+  "http://localhost:8001",
+  "http://localhost:8002",
+  "http://localhost:8004",
+];
+
+// PORT 8000 - FRONTEND
```
Comment (pzl111) [inline question-service/middlewares/cors.js:13]: I think frontend side is port 5173 for dev

**Pull request merged**